File: ios/WhoCalled/Views/Components.swift
import SwiftUI

/// Card with a subtle hairline border and no heavy shadow — clean, not flat-gray.
struct BorderedCard<Content: View>: View {
  var accent: Color? = nil
  @ViewBuilder var content: () -> Content

  var body: some View {
    content()
      .padding(16)
      .frame(maxWidth: .infinity, alignment: .leading)
      .background(Color(.systemBackground))
      .overlay(
        RoundedRectangle(cornerRadius: 12)
          .stroke((accent ?? WhoCalledColors.border).opacity(accent == nil ? 1 : 0.35), lineWidth: 1)
      )
      .clipShape(RoundedRectangle(cornerRadius: 12))
  }
}

enum BannerKind {
  case error, success, info

  var tint: Color {
    switch self {
    case .error: return WhoCalledColors.coral
    case .success: return WhoCalledColors.emerald
    case .info: return WhoCalledColors.indigo
    }
  }
  var icon: String {
    switch self {
    case .error: return "exclamationmark.triangle.fill"
    case .success: return "checkmark.circle.fill"
    case .info: return "info.circle.fill"
    }
  }
}

/// Bordered, lightly tinted feedback row with an SF Symbol — replaces bare red text.
struct FeedbackBanner: View {
  let kind: BannerKind
  let message: String

  var body: some View {
    HStack(spacing: 10) {
      Image(systemName: kind.icon).foregroundStyle(kind.tint)
      Text(message).font(.subheadline)
      Spacer(minLength: 0)
    }
    .padding(.horizontal, 12)
    .padding(.vertical, 10)
    .background(kind.tint.opacity(0.06))
    .overlay(
      RoundedRectangle(cornerRadius: 10).stroke(kind.tint.opacity(0.30), lineWidth: 1)
    )
    .clipShape(RoundedRectangle(cornerRadius: 10))
  }
}

/// Status pill (Bloqué / Alerté / Sûr) — tinted + bordered, no shadow.
struct StatusBadge: View {
  let label: String
  let color: Color
  let systemImage: String

  var body: some View {
    HStack(spacing: 4) {
      Image(systemName: systemImage).font(.caption)
      Text(label).font(.caption.weight(.semibold))
    }
    .foregroundStyle(color)
    .padding(.horizontal, 10)
    .padding(.vertical, 5)
    .background(color.opacity(0.08))
    .overlay(Capsule().stroke(color.opacity(0.35), lineWidth: 1))
    .clipShape(Capsule())
  }
}

/// Risk gauge: arc follows the internal score; center shows Faible / Modéré / Élevé.
/// Stroke and type scale with diameter so the compact detail size (≈72) never
/// clips "Modéré" over the arc.
struct ScoreGauge: View {
  let score: Int
  var status: String = "unknown"
  var diameter: CGFloat = 132

  private var color: Color {
    switch status {
    case "block": return WhoCalledColors.coral
    case "warn": return WhoCalledColors.amber
    case "allow": return WhoCalledColors.emerald
    default: return WhoCalledColors.blue
    }
  }

  private var riskLabel: String {
    switch status {
    case "block": "Élevé"
    case "warn": "Modéré"
    case "allow": "Faible"
    default: "—"
    }
  }

  private var stroke: CGFloat { max(4, min(12, diameter * 0.09)) }
  private var hole: CGFloat { max(18, diameter - (stroke + 3) * 2) }
  private var labelSize: CGFloat {
    let long = riskLabel.count >= 5
    switch diameter {
    case ..<56: return 8
    case ..<64: return long ? 9 : 10
    case ..<72: return long ? 10 : 11
    case ..<88: return long ? 12 : 13
    default: return 16
    }
  }
  private var captionSize: CGFloat { max(7, labelSize * 0.7) }
  private var showCaption: Bool { diameter >= 56 }

  var body: some View {
    ZStack {
      Circle()
        .trim(from: 0, to: 0.75)
        .stroke(Color.gray.opacity(0.15), style: StrokeStyle(lineWidth: stroke, lineCap: .round))
        .rotationEffect(.degrees(135))
      Circle()
        .trim(from: 0, to: 0.75 * CGFloat(min(max(score, 0), 100)) / 100)
        .stroke(color, style: StrokeStyle(lineWidth: stroke, lineCap: .round))
        .rotationEffect(.degrees(135))
        .animation(.easeOut(duration: 0.9), value: score)
      VStack(spacing: 1) {
        Text(riskLabel)
          .font(.system(size: labelSize, weight: .bold))
          .foregroundStyle(color)
          .minimumScaleFactor(0.6)
          .lineLimit(1)
          .frame(maxWidth: hole)
        if showCaption {
          Text("Risque")
            .font(.system(size: captionSize, weight: .medium))
            .foregroundStyle(WhoCalledColors.muted)
            .lineLimit(1)
        }
      }
      .frame(width: hole, height: hole)
    }
    .frame(width: diameter, height: diameter)
  }
}

/// Friendly empty-state placeholder. The glyph sits in a large soft tinted disc
/// (with a hairline ring) so an empty list reads as intentional and calm — never
/// broken or blank. An optional `action` slot offers the obvious next step.
struct EmptyState<Action: View>: View {
  let systemImage: String
  let title: String
  var subtitle: String? = nil
  var accent: Color = WhoCalledColors.indigo
  @ViewBuilder var action: () -> Action

  var body: some View {
    VStack(spacing: 10) {
      ZStack {
        Circle()
          .fill(accent.opacity(0.08))
          .frame(width: 88, height: 88)
        Circle()
          .stroke(accent.opacity(0.25), lineWidth: 1)
          .frame(width: 88, height: 88)
        Image(systemName: systemImage)
          .font(.system(size: 36))
          .foregroundStyle(accent.opacity(0.85))
      }
      Text(title).font(.headline).multilineTextAlignment(.center)
      if let subtitle {
        Text(subtitle)
          .font(.subheadline)
          .foregroundStyle(WhoCalledColors.muted)
          .multilineTextAlignment(.center)
      }
      action().padding(.top, 4)
    }
    .frame(maxWidth: .infinity)
    .padding(.horizontal, 24)
    .padding(.vertical, 36)
  }
}

extension EmptyState where Action == EmptyView {
  /// Convenience init for empty states without an action button.
  init(systemImage: String, title: String, subtitle: String? = nil,
       accent: Color = WhoCalledColors.indigo) {
    self.init(systemImage: systemImage, title: title, subtitle: subtitle,
              accent: accent, action: { EmptyView() })
  }
}

/// Brand lockup: shield mark in a blue disc + the "Who Called" wordmark. Used on
/// loading/branded screens so the identity is the name, not just a shield glyph.
struct BrandLogo: View {
  var horizontal = true
  var onDark = false

  private var mark: some View {
    ZStack {
      RoundedRectangle(cornerRadius: horizontal ? 12 : 18)
        .fill(onDark ? Color.white.opacity(0.16) : WhoCalledColors.blue)
        .frame(width: horizontal ? 44 : 64, height: horizontal ? 44 : 64)
      Image(systemName: "checkmark.shield.fill")
        .font(.system(size: horizontal ? 24 : 34))
        .foregroundStyle(.white)
    }
  }

  private var wordmark: some View {
    Text("Who Called")
      .font(.title2.bold())
      .foregroundStyle(onDark ? Color.white : WhoCalledColors.nightBlue)
  }

  var body: some View {
    if horizontal {
      HStack(spacing: 12) { mark; wordmark }
    } else {
      VStack(spacing: 12) { mark; wordmark }
    }
  }
}

/// Full-screen branded loader: the Who Called lockup over a spinner.
struct BrandedLoader: View {
  var body: some View {
    VStack(spacing: 20) {
      Spacer()
      BrandLogo(horizontal: false)
      ProgressView().tint(WhoCalledColors.blue)
      Spacer()
    }
    .frame(maxWidth: .infinity)
    .padding(32)
  }
}

/// Vivid blue → navy gradient header band.
struct GradientHeader: View {
  let title: String
  var subtitle: String? = nil

  var body: some View {
    VStack(alignment: .leading, spacing: 4) {
      Text(title).font(.title2.bold()).foregroundStyle(.white)
      if let subtitle {
        Text(subtitle).font(.subheadline).foregroundStyle(.white.opacity(0.85))
      }
    }
    .frame(maxWidth: .infinity, alignment: .leading)
    .padding(20)
    .background(
      // Gentle same-family gradient (vivid blue → slightly darker blue),
      // never fading toward navy/grey.
      LinearGradient(
        colors: [WhoCalledColors.blueDark, WhoCalledColors.blue],
        startPoint: .topLeading, endPoint: .bottomTrailing)
    )
    .clipShape(RoundedRectangle(cornerRadius: 18))
  }
}

File: android/app/src/main/java/com/whocalled/android/ui/components/Components.kt
package com.whocalled.android.ui.components

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.expandVertically
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.shrinkVertically
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.animation.scaleIn
import androidx.compose.animation.scaleOut
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.foundation.lazy.LazyListState
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.CheckCircle
import androidx.compose.material.icons.rounded.ErrorOutline
import androidx.compose.material.icons.rounded.Info
import androidx.compose.material.icons.rounded.KeyboardArrowUp
import androidx.compose.material.icons.rounded.VerifiedUser
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.SmallFloatingActionButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.whocalled.android.ui.theme.WCColor
import kotlinx.coroutines.launch

/** Card with a subtle hairline border and no shadow — clean, not heavy. */
@Composable
fun BorderedCard(
    modifier: Modifier = Modifier,
    accent: Color? = null,
    content: @Composable ColumnScope.() -> Unit,
) {
    Column(
        modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(12.dp))
            .background(MaterialTheme.colorScheme.surface)
            .border(
                BorderStroke(1.dp, accent?.copy(alpha = 0.35f) ?: MaterialTheme.colorScheme.outline),
                RoundedCornerShape(12.dp),
            )
            .padding(16.dp),
        content = content,
    )
}

enum class BannerKind { Error, Success, Info }

/**
 * Feedback banner — a bordered, lightly tinted row with an icon. No gray fill,
 * no shadow: structure comes from a hairline border and a colored left rail.
 */
@Composable
fun FeedbackBanner(kind: BannerKind, message: String, modifier: Modifier = Modifier) {
    val (tint, icon: ImageVector) = when (kind) {
        BannerKind.Error -> WCColor.Coral to Icons.Rounded.ErrorOutline
        BannerKind.Success -> WCColor.Emerald to Icons.Rounded.CheckCircle
        BannerKind.Info -> WCColor.Indigo to Icons.Rounded.Info
    }
    Row(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(10.dp))
            .background(tint.copy(alpha = 0.06f))
            .border(BorderStroke(1.dp, tint.copy(alpha = 0.30f)), RoundedCornerShape(10.dp))
            .padding(horizontal = 12.dp, vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Icon(icon, contentDescription = null, tint = tint, modifier = Modifier.size(20.dp))
        Text(message, style = MaterialTheme.typography.bodyMedium)
    }
}

@Composable
fun AnimatedBanner(kind: BannerKind, message: String?, modifier: Modifier = Modifier) {
    AnimatedVisibility(
        visible = message != null,
        enter = fadeIn() + expandVertically(),
        exit = fadeOut() + shrinkVertically(),
    ) {
        FeedbackBanner(kind, message.orEmpty(), modifier)
    }
}

/** Status pill (Bloqué / Alerté / Sûr) — tinted, bordered, no shadow. */
@Composable
fun StatusBadge(label: String, color: Color, icon: ImageVector, modifier: Modifier = Modifier) {
    Row(
        modifier = modifier
            .clip(RoundedCornerShape(50))
            .background(color.copy(alpha = 0.08f))
            .border(BorderStroke(1.dp, color.copy(alpha = 0.35f)), RoundedCornerShape(50))
            .padding(horizontal = 10.dp, vertical = 5.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Icon(icon, contentDescription = null, tint = color, modifier = Modifier.size(16.dp))
        Text(label, style = MaterialTheme.typography.labelMedium, color = color, fontWeight = FontWeight.SemiBold)
    }
}

/**
 * Icon-only variant of [StatusBadge] — same tinted disc, no label. Used in dense
 * lists (the calls history) where a worded pill ("Non évalué", "Indésirable")
 * would eat the row's width and push the number onto a second line. The label
 * still ships as the content description, so the status stays available to
 * TalkBack and reads out exactly like the full badge.
 */
@Composable
fun StatusDot(label: String, color: Color, icon: ImageVector, modifier: Modifier = Modifier) {
    Box(
        modifier = modifier
            .size(30.dp)
            .clip(RoundedCornerShape(50))
            .background(color.copy(alpha = 0.08f))
            .border(BorderStroke(1.dp, color.copy(alpha = 0.35f)), RoundedCornerShape(50)),
        contentAlignment = Alignment.Center,
    ) {
        Icon(icon, contentDescription = label, tint = color, modifier = Modifier.size(17.dp))
    }
}

/** Vivid blue → navy gradient header band; modest rounding at the bottom. */
@Composable
fun GradientHeader(
    title: String,
    subtitle: String? = null,
    modifier: Modifier = Modifier,
    leading: (@Composable () -> Unit)? = null,
    trailing: (@Composable () -> Unit)? = null,
) {
    Box(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(bottomStart = 16.dp, bottomEnd = 16.dp))
            .background(
                androidx.compose.ui.graphics.Brush.linearGradient(
                    colors = listOf(WCColor.BlueDark, WCColor.Blue),
                ),
            )
            .padding(horizontal = 16.dp, vertical = 12.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            if (leading != null) {
                leading()
            }
            Column(Modifier.weight(1f)) {
                Text(title, style = MaterialTheme.typography.titleMedium, color = Color.White, fontWeight = FontWeight.Bold, maxLines = 1)
                if (subtitle != null) {
                    Text(subtitle, style = MaterialTheme.typography.bodySmall, color = Color.White.copy(alpha = 0.85f), maxLines = 1)
                }
            }
            if (trailing != null) trailing()
        }
    }
}

/**
 * Friendly empty-state placeholder. Rather than a bare icon, the glyph sits in a
 * large soft tinted disc (with a hairline ring) so an empty list reads as
 * intentional and calm — never broken or blank. Used wherever a list can
 * legitimately be empty (no calls, no reports, no trends yet). An optional
 * [action] slot lets the screen offer the obvious next step (e.g. "Mettre à jour").
 */
@Composable
fun EmptyState(
    icon: ImageVector,
    title: String,
    subtitle: String? = null,
    modifier: Modifier = Modifier,
    accent: Color = WCColor.Indigo,
    action: (@Composable () -> Unit)? = null,
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 24.dp, vertical = 36.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Box(
            modifier = Modifier
                .size(88.dp)
                .clip(RoundedCornerShape(50))
                .background(accent.copy(alpha = 0.08f))
                .border(BorderStroke(1.dp, accent.copy(alpha = 0.25f)), RoundedCornerShape(50)),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                icon,
                contentDescription = null,
                tint = accent.copy(alpha = 0.85f),
                modifier = Modifier.size(40.dp),
            )
        }
        Text(
            title,
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.SemiBold,
            color = MaterialTheme.colorScheme.onSurface,
            textAlign = androidx.compose.ui.text.style.TextAlign.Center,
        )
        if (subtitle != null) {
            Text(
                subtitle,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = androidx.compose.ui.text.style.TextAlign.Center,
            )
        }
        if (action != null) {
            Box(Modifier.padding(top = 4.dp)) { action() }
        }
    }
}

/**
 * Brand lockup: shield mark in a blue disc + the "Who Called" wordmark. Used on
 * loading/branded screens so the identity is the name, not just a shield glyph.
 */
@Composable
fun BrandLogo(
    modifier: Modifier = Modifier,
    horizontal: Boolean = true,
    onDark: Boolean = false,
) {
    val markBg = if (onDark) Color.White.copy(alpha = 0.16f) else WCColor.Blue
    val markTint = Color.White
    val textColor = if (onDark) Color.White else WCColor.NightBlue
    val mark = @Composable {
        Box(
            Modifier
                .size(if (horizontal) 44.dp else 64.dp)
                .clip(RoundedCornerShape(if (horizontal) 12.dp else 18.dp))
                .background(markBg),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                Icons.Rounded.VerifiedUser,
                contentDescription = null,
                tint = markTint,
                modifier = Modifier.size(if (horizontal) 26.dp else 38.dp),
            )
        }
    }
    val wordmark = @Composable {
        Text(
            "Who Called",
            style = MaterialTheme.typography.titleLarge,
            fontWeight = FontWeight.Bold,
            color = textColor,
        )
    }
    if (horizontal) {
        Row(
            modifier,
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(12.dp),
        ) { mark(); wordmark() }
    } else {
        Column(
            modifier,
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) { mark(); wordmark() }
    }
}

/** Full-screen branded loader: the Who Called lockup over a spinner. */
@Composable
fun BrandedLoader(modifier: Modifier = Modifier) {
    Column(
        modifier.fillMaxSize().padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(20.dp),
    ) {
        Box(Modifier.weight(1f))
        BrandLogo(horizontal = false)
        androidx.compose.material3.CircularProgressIndicator(
            color = WCColor.Blue,
            strokeWidth = 3.dp,
        )
        Box(Modifier.weight(1f))
    }
}

/**
 * A screen whose [header] is pinned at the top — exactly as it looks at rest,
 * with no shrinking or transformation — while only the list below it scrolls.
 * A floating "remonter" button appears once the user scrolls down and animates
 * the list back to the top on tap (the requested UX hint).
 */
@Composable
fun ScrollableScreen(
    header: @Composable () -> Unit,
    modifier: Modifier = Modifier,
    content: LazyListScope.() -> Unit,
) {
    val listState = rememberLazyListState()
    val scope = rememberCoroutineScope()
    val scrolled by remember {
        derivedStateOf {
            listState.firstVisibleItemIndex > 0 || listState.firstVisibleItemScrollOffset > 200
        }
    }

    Box(modifier.fillMaxSize()) {
        Column(Modifier.fillMaxSize()) {
            // Fixed header — outside the scroll, always shown as-is.
            header()
            // Only this list scrolls, underneath the pinned header.
            LazyColumn(
                state = listState,
                modifier = Modifier.fillMaxSize(),
                contentPadding = PaddingValues(top = 12.dp, bottom = 32.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp),
                content = content,
            )
        }

        // Floating scroll-to-top button (the UX hint to get back up).
        AnimatedVisibility(
            visible = scrolled,
            enter = scaleIn() + fadeIn(),
            exit = scaleOut() + fadeOut(),
            modifier = Modifier
                .align(Alignment.BottomEnd)
                .padding(20.dp),
        ) {
            SmallFloatingActionButton(
                onClick = { scope.launch { listState.animateScrollToItem(0) } },
                containerColor = WCColor.Indigo,
                contentColor = Color.White,
            ) {
                Icon(Icons.Rounded.KeyboardArrowUp, contentDescription = "Remonter en haut")
            }
        }
    }
}
